lib/oauth2-client.js

```diff
@@ -0,0 +1,107 @@
+/*!
+ * Copyright (c) 2021 Digital Bazaar, Inc. All rights reserved.
+ */
+import * as storage from './storage.js';
+import * as bedrock from 'bedrock';
+import {httpClient} from '@digitalbazaar/http-client';
+import logger from './logger.js';
+const {util: {BedrockError}} = bedrock;
+
+export class OAuth2Client {
+  /**
+   * Constructs an OAuth2Client instance to return an access token. It first
+   * checks the database to see if an access token already exists for the client
+   * id; if not, it creates an access token then stores it in the database.
+   *
+   * @param {object} options - Options to use.
+   * @param {string} options.config - The "oauth2-client-grant" properties.
+   * @param {object} options.httpsAgent - A NodeJS HTTPS Agent (`https.Agent`).
+   *
+   * @returns {OAuth2Client} The `OAuth2Client` instance.
+   */
+  constructor({config, httpsAgent} = {}) {
+    this.client_id = config.client_id;
+    this.client_secret = config.client_secret;
+    this.token_endpoint = config.token_endpoint;
+    this.grant_type = config.grant_type;
+    this.scope = config.scope;
+    this.maxRetries = config.maxRetries;
+    this.httpsAgent = httpsAgent;
+  }
+
+  /**
+   * Gets a new access token by first checking the database to see if an
+   *  access token already exists for the client id; if not it creates an
+   *  access token then stores it in the database.
+   *
+   * @returns {string} The access token.
+   */
+  async getAccessToken() {
+    const {
+      client_id, client_secret, token_endpoint, grant_type, scope, httpsAgent,
+      maxRetries
+    } = this;
+    let {token} = await storage.get({client_id}) || {};
+    if(!(token && token.access_token)) {
+      token = await getNewAccessToken({
+        client_id, client_secret, token_endpoint, grant_type, scope, httpsAgent,
+        maxRetries
+      });
+      await storage.insert({token});
+    }
```

Comment:
So this needs to be wrapped using lru-memoize so that when a token isn't available in the database, multiple calls aren't made to get a new one. We've used that pattern many times, one example to look at is here:

https://github.com/digitalbazaar/bedrock-kms/blob/main/lib/keystores.js#L188-L189

I believe we'd want to create a helper function here that does all the things I highlighted and that has some configurable max age set on it with a sane default.

Reply:
Added! What is a "sane default"? I started with 10, but I'm not sure what the expected ballpark is.

Also, what should the maxAge be? I set it to 2 weeks, but I think that might not be correct since there could be a few second overlap between the token expiring but still being in the cache.

Reply:
@mandyvenables I think 2 weeks is fine for now.